core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -555,12 +555,15 @@ private[spark] class BlockManager(
 
   /**
    * Return a list of locations for the given block, prioritizing the local machine since
-   * multiple block managers can share the same host.
+   * multiple block managers can share the same host, followed by hosts on the same rack.
    */
   private def getLocations(blockId: BlockId): Seq[BlockManagerId] = {
     val locs = Random.shuffle(master.getLocations(blockId))
     val (preferredLocs, otherLocs) = locs.partition { loc => blockManagerId.host == loc.host }
-    preferredLocs ++ otherLocs
+    val (sameRackLocs, differentRackLocs) = otherLocs.partition {
+      loc => blockManagerId.topologyInfo == loc.topologyInfo
```

Comment:
If `blockManagerId.topologyInfo` is `None`, we will prefer the locations with empty `topologyInfo`. It is slightly different with what the shuffling wants to do here.